packages/flutter/test/widgets/slivers_keepalive_test.dart

```diff
@@ -0,0 +1,564 @@
+
+  testWidgets('Sliver list should not dispose widget with key during in screen reordering', (WidgetTester tester) async {
+    final Map<String, bool> childOneOptions = <String, bool> {
+      'keepAlive': true,
+      'shouldDispose': false,
+    };
+    final Map<String, bool> childTwoOptions = <String, bool> {
+      'keepAlive': false,
+      'shouldDispose': false,
+    };
+    final Map<String, bool> childThreeOptions = <String, bool> {
+      'keepAlive': true,
+      'shouldDispose': true,
+    };
+    final List<Widget> childList= <Widget>[
+      WidgetTest1('child 1', key: UniqueKey(), optionsRef: childOneOptions),
+      WidgetTest2('child 2', key: UniqueKey(), optionsRef: childTwoOptions),
+      WidgetTest3('child 3', optionsRef: childThreeOptions),
+    ];
+    final GlobalKey widgetKey = GlobalKey<_SwitchingChildListTest>();
+    await tester.pumpWidget(
+        SwitchingChildListTest(childList, key: widgetKey, viewportFraction: 0.1)
+    );
+    expect(find.text('child 1', skipOffstage: false), findsOneWidget);
+    expect(find.text('child 2', skipOffstage: false), findsOneWidget);
+    expect(find.text('child 3', skipOffstage: false), findsOneWidget);
+    final _SwitchingChildListTest state = widgetKey.currentState;
+
+    state.switchChildrenOrder(0, 2);
+    await tester.pump();
+
+    state.switchChildrenOrder(0, 1);
+    await tester.pump();
+
+    state.switchChildrenOrder(1, 2);
+    await tester.pump();
+
+    state.switchChildrenOrder(0, 1);
+    await tester.pump();
+
+    childOneOptions['shouldDispose'] = true;
+    childTwoOptions['shouldDispose'] = true;
+  });
+
+}
+
+class SwitchingChildBuilderTest extends StatefulWidget {
+  SwitchingChildBuilderTest(this.children, {
+    Key key}) : super(key: key);
+
+  final List<Widget> children;
+
+  @override
+  _SwitchingChildBuilderTest createState() {
+    return _SwitchingChildBuilderTest();
+  }
+}
+
+class _SwitchingChildBuilderTest extends State<SwitchingChildBuilderTest> {
+  List<Widget> children;
+  Map<Key, int> _mapKeyToIndex;
+
+  void switchChildrenOrder(int i, int j) {
+    setState(() {
+      final Widget w = children[i];
+      children[i] = children[j];
+      children[j] = w;
+      if (children[i].key != null)
+        _mapKeyToIndex[children[i].key] = i;
+      if (children[j].key != null)
+        _mapKeyToIndex[children[j].key] = j;
+    });
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    children = widget.children;
+    _mapKeyToIndex = <Key, int>{};
+    children.asMap().forEach((int index , Widget w) {
+      if (w.key != null) {
+        _mapKeyToIndex[w.key] = index;
+      }
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Directionality(
+      textDirection: TextDirection.ltr,
+      child: Center(
+        child: Container(
+          height: 100,
+          child: CustomScrollView(
+            cacheExtent: 0,
+            slivers: <Widget>[
+              SliverFillViewport(
```

Comment:
I would be good to also have a couple of tests that use a plain-old SliverList (e.g. ListView) or SliverFixedExtent list.

Reply:
added